man/pidfile.8
.TH ninit\-pidfile 8 "Dec 28, 2009"
.SH NAME
ninit\-pidfile \- work around daemons that always fork
.SH SYNOPSIS
.B ninit\-pidfile
.I servicename
.I pidfile
[
.I -H home
]
.I command
.I [parameters]

.SH DESCRIPTION
.B ninit\-pidfile
is used to work around daemons that insist on forking into the background,
but that do write a correct pid file.

.B ninit\-pidfile 
forks the actual service, then waits for the pidfile to
be written. Once it can read the pid from the pidfile it will tell
ninit the real pid and quit.

.SH OPTIONS
.TP
.I servicename
the name of the service
ninit\-pidfile is installed for
.TP
.I pidfile
the filename to read the pid from
.TP
.I command
the real command to start
.TP
.I parameters
additional parameters for the command
.TP
.B \-H \fIhome\fR
.br
the home of ninit.  Default: /etc/ninit

.SH USAGE
With
.B ninit 
you can prepare a service and if it forks and writes
the PID in some file then type in the service directory:

	echo /path/to/deamon.pidfile > 
.I pidfile

For apache this looks like:

	cd /etc/ninit/apache
.br
	echo /var/run/apache.pid > 
.I pidfile

Don't use hard or soft links here!  
.br
Then 
.B ninit 
will start apache using 
.B ninit\-pidfile

.SH AUTHOR
.B pidfilehack
was written 
by Felix von Leitner.

This manpage was written by Erich Schubert <erich@debian.org>
for the Debian GNU/Linux operating system.

.SH "SEE ALSO"
ninit(8), nsvc(8), pidfilehack(8)


CHANGES
0.14:   2010-01-19 00:27:01
  added nkillall.8  
  gid is not default anymore.  see ninitfeatures.h
  uid supports 22:33:100:200:300
  nkillall: \e escape
  shutdown: set cron off
  nkillall: signals with letters like nsvc
  shutdown and nkillall: makes more sync(2), flag -q
  added THANKS ;-)
  added helpers: remove procfs;  thanks to them snvc is smaller
  nsvc options -onumber -unumber; option -E
  Cron has option -C
  /etc/ninit/.nsvc_help
  nsvc, shutdown, reload uses variables NINIT_MEMORY and NINIT_HOME
  nsvc: option -S
  shutdown: option -s
  manpages for all programs in /sbin ;-)
  update manpages (Boris Grozev)
  Makefile: man_txt
  serdo.c: read only one file again
  better sleep in open_inout.h
  serdo.c improvments; buildin: killall5, simple echo
  nsvc.c: -C+number
  scripts/conf
  ninit.8: NONROOT USAGE, using /sbin/init with ninit
  ninit-scan: (Wayne Marshall)

0.13.7:   2009-12-16 10:47:50
  bootlog.c (flush_root)
  0.13.1.4 renamed to 0.13.2
  service start 'end' (if it exixts; X_OK) after finishing
  fixed a bug in shutdown.c (thx to Jan)
  With flag -s ninit-shutdown don't start halt and reboot services
  shutdown.c:  improvements; flags -S -q
  shutdown.c:  removed flags -s -q; new flags -E -v -T
  added nkillall
  shutdown.c:  options -m and -E work together
  nkillall print escape like echo

0.13.1:   2009-06-27 12:59:38
  updated bootlog
  Makefile logs 'make tests' in tests_log using boolog
  fixing quotes in serdo

0.13:   2009-01-10 18:11:42
  it's possible to set parameters in wait.  example: some_service:180:3
  install-bin print help and string errors
  better sync mode.  maxsync is removed.  echo 200 > sync
  updated scripts/ninit_test.sh
  comments (#) in depends, params, environ, wait, softlimit
  install-bin (verbose mode)
  nsvc -V
  errmsg_put.c; buffer_*.c
  setup, rsetup, sys-rsetup starts with args: $1=service $2=service_pid
  EXTRACT_* trick (automatic create header files)
  stuct process is 20 bytes on x86_64 (see struct_root.h)
  changed BIFFER_INIT macro
  utmp_io.c
  Makefile: install_other
  serdo is intslled only with: make install_other
  t_write.h uses PIPE_BUF (thx Laurent Bercot)
  nsvc.8 improvements
  ninit_test.sh creates services in ./etc/ninit/
  ninit-huge
  added printf.c and ninit.spec
  make FLAG_DEBUG=no
  removed flag X_OK in some syscalls access.
  better addprocess

0.12.1:   2008-01-03 17:15:08
  ninit-depends.  convert directory to file (depends.dir -> depends)
  removed unused stat.h in headers
  updated serdo (static)
  two environ vars: NINIT_HOME, INIT_HOME.  (Thanks to Stamatis Mitrofanis)
  ninit logo: /etc/ninit/.sync
  install-bin uses chown32 instead chown on some systems
  err.c and err_b.c uses the macros va_start and va_arg
  fu.c (u &= 0xffffffff)

0.12:
  almost all arrays are moved in stack.  small data and bss sections.
  assembler functions (i386) for some DJB functions.
  modifications in Makefile.
  100 Euro security guarantee.

0.11.2:
  ninit-reload accept agument -e (change environ) -E file_env
  program ~/sys/update
  service flags: pause, pause-wait
  service flag cron:  a:b[:c]
    can contain many lines.
  flag -K for nsvc.
  pidfilehack --> ninit-pidfile.
  make i386 -- build static daemons (don't use dietlibc)
  Makefile is smaller and simpler
  gcc-4.1.2 works now.
  install-bin
  stuct mem in initialized in the main program.

0.11.1:
  ninit-reload accepts args (-r3 -r5 -r33 ...)
  ninit-inittab sets the variables NINIT_RUNLEVEL, INIT_VERSION
  if /etc/minit/{in|out} FIFO's exist then make symbolic links.
  new run flags: sys-rsetup, pidfile
  if the service sysvinit exist, ninit create /dev/initctl 
  ninit-reload and nsvc checks for environ NINIT_HOME
  nsvc -[drR] ALL - change all services.
    Change service ALL with: nsvc -[drR] ops ALL
  new program: ninit-sysvinit.  configuration file: sysvinit-timeout
    fimeout[:fork-mode].  example:  echo 600:0 > sysvinit-timeout
  ninit catch signal SIGPWR
  new run flags: alarm, maxsync
  nsvc output is different
  shell script ninit-service
  new program run-wait

0.11:
  Sun Jun 17 22:08:22 EEST 2007
  I decided to rename all without changes.
  The name ninit must be spelled nano-init.

0.9.11:
  rewritten by Nikola Vladov

0.9.1:
  fix embarassing typo in msvc (Gelu G. Lupas)

0.9:
  See minit CHANGES - http://www.fefe.de/minit


pidfile.c
#include <unistd.h>
#include <errno.h>
#include <alloca.h>
#include <fcntl.h>
#include "ninitfeatures.h"
#include "process_defs.h"

static int infd, outfd;

#include "open_inout.h"
#include "tryservice.h"

int main(int argc, char* argv[]) {
  char *x, *initroot, **arg;
  struct process pr[6];
  unsigned long pidfile;
  int count=0, pid, ret=1;

  if (argc<4) die(1,
	"usage: ninit-pidfile",
	" service /var/run/daemon.pid [-H home] /sbin/daemon ...");
  errmsg_iam("ninit-pidfile");

  if (unlink(argv[2]) && errno != ENOENT)
    die(1, "could not remove pid file: ", argv[2]);

  arg = argv + 3;
  x = arg[0];

  if (argc > 5 && x[0] =='-' && x[1] =='H' && x[2] ==0) {
    initroot = arg[1];
    arg += 2;
  } else
    initroot = INITROOT;

  while ((pid = fork()) < 0) nano_sleep(0, 500000000); /* 0.5 sec */

  if (pid == 0) {
    x = arg[0];

    arg[0] = x + str_rchr(x,'/');
    if (arg[0]) arg[0]++;
    else        arg[0]=x;

    execve(x,arg,environ);
    die(3, "execve failed on: ", x);
  }

  while (1) {
    if (read_ulongs(argv[2], &pidfile, 1)) {
      if (str_len(argv[1]) > 320) ret=1;
      else {
	ret = tryservice(initroot,argv[1],"P",fu(pidfile), pr);
	ret = (ret != sizeof(pr[0]) || pr->pid != (pid_t)pidfile);
      }
      break;
    }
    if (++count>=30) break;
    else
      nano_sleep(1, 0);
  }

  if (ret) carp("could not set PID of service ", argv[1]);
  return ret;
}
